### ISSUE TITLE
✨ Add MH-PatchCore as a standalone anomaly detection model

### ISSUE THREAD

**Issue by NickF93** [open-edge-platform/anomalib#3708]
### Problem Description

Anomalib currently provides `Patchcore`, which uses Euclidean nearest-neighbor retrieval and constructs its coreset after collecting the normal patch embeddings.

I would like to contribute **Mahalanobis PatchCore (MH-PatchCore)**, a covariance-aware and streaming-compatible PatchCore variant. It addresses two limitations of the standard formulation:

1. Euclidean retrieval treats embedding dimensions as uncorrelated and equally scaled.
2. Offline memory-bank construction materializes the full normal patch pool before coreset selection, which can lead to high peak RAM usage.

MH-PatchCore estimates a regularized covariance model in a reduced embedding space and whitens the embeddings before nearest-neighbor retrieval. Euclidean retrieval in the whitened space is therefore equivalent to Mahalanobis retrieval. Its multi-pass but bounded-memory fitting procedure uses incremental dimensionality reduction, online covariance estimation, and merge-reduce memory-bank construction without storing all normal patches simultaneously.

The method remains a one-class anomaly detection and localization model. “Streaming-compatible” refers to bounded-memory fitting over a re-iterable normal dataset, not continual post-deployment adaptation.

### References and ownership

- **Preprint:** [Mahalanobis PatchCore: Covariance-Aware and Streaming-Compatible Industrial Anomaly Detection](https://arxiv.org/abs/2605.27748)
- **Publication status:** preprint available on arXiv; the manuscript is currently under review at *Engineering Applications of Artificial Intelligence (EAAI)*
- **Official implementation:** [NickF93/MH-PatchCore](https://github.com/NickF93/MH-PatchCore)
- **Authors:** Niccolò Ferrari, Oligert Osmani, and Evelina Lamma
- I am Niccolò Ferrari, the first and corresponding author and one of the upstream code copyright holders.
- **Upstream code license:** MIT, copyright © 2025 Niccolò Ferrari and Oligert Osmani

### Proposed Solution

I propose integrating MH-PatchCore as a separate public model, tentatively named `MHPatchcore`, implemented with `MemoryBankMixin` and `AnomalibModule`.

I prefer a standalone model because covariance fitting, incremental reduction, whitening state, and bounded-memory bank construction change the fitting lifecycle and serialized model state, rather than only changing PatchCore's final distance calculation. However, I would like maintainers to confirm whether they prefer:

1. a separate `MHPatchcore` model; or
2. an extension/configuration of the existing `Patchcore` model.

Where possible, the port would reuse anomalib's existing PatchCore feature extraction, patch alignment, anomaly-map generation, and nearest-neighbor components.

### Proposed initial API

The exact naming and parameters are open to maintainer feedback. An illustrative constructor is:

```python
from anomalib.models import MHPatchcore

model = MHPatchcore(
    backbone="wide_resnet50_2",
    layers=("layer2", "layer3"),
    pre_trained=True,
    pca_variance_ratio=0.99,
    covariance_shrinkage=0.07,
    memory_bank_size=1000,
    local_coreset_size=256,
    num_neighbors=9,
)
```

CLI usage would follow the existing model interface:

```bash
anomalib train \
    --model MHPatchcore \
    --data anomalib.data.MVTecAD \
    --data.category hazelnut
```

Proposed integration behavior:

- **Learning type:** `LearningType.ONE_CLASS`
- **Training input:** normal images through anomalib's standard typed `Batch`
- **Fitting:** incremental reducer fit, online covariance estimation, whitening factorization, and bounded-memory memory-bank construction
- **Inference output:** standard `pred_score` and pixel-level `anomaly_map`
- **Checkpoint state:** reducer state, covariance mean/factorization, and final memory bank
- **Pre/post-processing:** standard anomalib `PreProcessor` and `PostProcessor`
- **Evaluation and visualization:** standard `Evaluator` and `Visualizer`
- **CLI/config:** explicit, jsonargparse-compatible constructor parameters

A key architecture question is how anomalib wants a model to perform the required re-iterable, multi-pass fit without bypassing the normal Engine/Lightning lifecycle. I would implement this only after maintainers approve the approach.

### Initial contribution scope

The first PR would include the canonical MH-PatchCore configuration:

- WideResNet-50-2 feature extraction from `layer2` and `layer3`
- incremental PCA
- online covariance estimation
- fixed covariance shrinkage and stable Cholesky whitening
- merge-reduce k-center memory-bank construction
- PatchCore-style reweighted scoring
- image-level scores and pixel-level anomaly maps
- CPU and accelerator-compatible PyTorch implementation where supported

The following would be out of scope for the initial PR:

- the upstream plugin/configuration framework
- GeoReS and other paper ablations
- mini-batch k-means alternatives
- the standalone upstream CLI and benchmark runner
- private industrial datasets
- continual or post-deployment online adaptation

### Extend the existing `Patchcore` class

Covariance-aware distance could be introduced through parameters such as `distance="mahalanobis"` and `memory_bank_strategy="streaming"`.

This could reduce duplicated PatchCore code, but it would also add incremental reducer state, covariance state, multiple fitting passes, and a different bank-construction lifecycle to the existing model. My preference is therefore a separate model that reuses shared PatchCore components. I welcome the maintainers' architectural decision.

### Add only reusable components

The incremental reducer, covariance estimator, and bounded-memory coreset could be contributed as generic components without exposing MH-PatchCore as a model. This would not provide users with the complete published method or its standard CLI/config integration.

### Keep the implementation external

The upstream package is usable independently, but it does not provide anomalib's unified data modules, Engine, evaluation, visualization, export, and configuration interfaces.

### Paper-reported benchmark results

These are results reported by the authors using the upstream implementation, **not yet an anomalib-port reproduction**.

Canonical MVTec AD protocol:

- 15 categories, one model per category
- fixed seed (`42` in the released baseline configuration)
- WideResNet-50-2 pretrained on ImageNet-1k
- features from `layer2` and `layer3`
- resize to 256 px and center crop to 224×224
- no training augmentation
- incremental PCA retaining 99% explained variance
- covariance shrinkage `0.07`
- final bank size `K=1000`
- local merge-reduce coreset size `256`
- 9 neighbors for image-score reweighting

| Method | Mean image AUROC | Mean pixel AUROC | Mean peak RAM | Maximum peak RAM | Inference |
|---|---:|---:|---:|---:|---:|
| Original offline PatchCore | 0.991 | 0.980 | 4.125 GB | 5.412 GB | 467.092 ms/image |
| Streaming MH-PatchCore | 0.989 | 0.978 | 2.430 GB | 2.775 GB | 42.610 ms/image |

Inference time is the image-weighted mean across the MVTec AD test images.

Reported MVTec experiments were CPU-only on Linux using:

- Intel Core i7-9750H
- 12 CPU threads
- 15 GiB RAM and 2 GiB swap
- Python 3.11.11
- PyTorch 2.10.0

GPU acceleration was not required for the reported experiments.

After architecture approval, I will reproduce the accepted configuration through anomalib and report the anomalib results separately from the paper results, including per-category metrics, fit time, inference time, and peak RAM.

### Licensing and provenance

The upstream MH-PatchCore code is MIT licensed. An anomalib contribution would be made under anomalib's Apache-2.0 contribution terms while preserving the upstream MIT copyright and notices where code is adapted.

One upstream utility, `mhpc/util/patcher.py`, states that it was adapted from Amazon's Apache-2.0 `patchcore-inspection` implementation. My proposed port would reuse anomalib's existing PatchCore utilities instead of copying that file.

If any Amazon-derived implementation remains necessary, I will preserve its Apache-2.0 attribution and notices in addition to:

- adding the required colocated `LICENSE`;
- updating `third-party-programs.txt`; and
- preserving the MIT attribution to Niccolò Ferrari and Oligert Osmani.

### Planned PR deliverables

After architectural approval, the model PR would include:

- `AnomalibModule`/`MemoryBankMixin` integration
- public model exports and discovery
- CLI/config example
- unit and integration tests
- numerical tests for covariance estimation and whitening
- checkpoint round-trip tests
- tests that do not require network access
- model README and reference documentation
- measured anomalib benchmarks
- sample anomaly maps
- third-party licensing records
- changelog entry

I will wait for explicit maintainer approval of the standalone-model decision and multi-pass fitting architecture before starting the full port.

**Comment by rajeshgangireddy** [comment]
Hi @NickF93 
Thanks for the proposal. 
please do contribute your work on mH-Patchcore to anomalib. 

> I propose integrating MH-PatchCore as a separate public model, tentatively named MHPatchcore, implemented with MemoryBankMixin and AnomalibModule.

Sounds good



About the API 


```
model = MHPatchcore(
    backbone="wide_resnet50_2",
    layers=("layer2", "layer3"),
    pre_trained=True,
    pca_variance_ratio=0.99,
    covariance_shrinkage=0.07,
    memory_bank_size=1000,
    local_coreset_size=256,
    num_neighbors=9,
)
```

This looks okay. Nitpick : If possible, would be great if you can remove exposing the parameters that won't matter a lot to keep the API simpler. But totally fine as is, as well. 
Please do also add some results with anomalib and the numbers from your paper in src/anomalib/models/image/<new_model>/README.md

I will assign it to you. 
Let us know if you need any help.
